metronome/networking/src/io/iohk/metronome/networking/ConnectionHandler.scala

```diff
@@ -0,0 +1,215 @@
+    * @param key, key of remote peer
+    */
+  def getConnection(key: K): F[Option[HandledConnection[F, K, M]]] =
+    connectionsRegister.getConnection(key)
+
+  private def callCallBackIfNotClosed(
+      handledConnection: HandledConnection[F, K, M]
+  ): F[Unit] = {
+    cancelToken.tryGet.flatMap {
+      case Some(_) => Concurrent[F].unit
+      case None    => connectionFinishCallback(handledConnection)
+    }
+  }
+
+  /** Connections multiplexer, it receives both incoming and outgoing connections and start reading incoming messages from
+    * them concurrently, putting them on received messages queue.
+    * In case of error or stream finish it cleans up all resources.
+    */
+  private def handleConnections: F[Unit] = {
+    Iterant
+      .repeatEvalF(connectionQueue.poll)
+      .mapEval { connection =>
+        Concurrent[F].delay(numberOfRunningConnections.increment()).flatMap {
+          _ =>
+            Iterant
+              .repeatEvalF(
+                withCancelToken(cancelToken, connection.incomingMessage)
+              )
+              .takeWhile(_.isDefined)
+              .map(_.get)
+              .mapEval {
+                case Right(m) =>
+                  messageQueue.offer(
+                    MessageReceived(connection.key, m)
+                  )
+                case Left(e) =>
+                  Concurrent[F].raiseError[Unit](
+                    UnexpectedConnectionError(e, connection.key)
+                  )
+              }
+              .guarantee(
+                closeAndDeregisterConnection(connection)
+                  .flatMap(_ => callCallBackIfNotClosed(connection))
+              )
+              .completedL
+              .start
+        }
+      }
+      .completedL
+  }
+
+  // for now shutdown of all connections is completed in background
+  private def shutdown: F[Unit] = cancelToken.complete(()).attempt.void
+}
+
+object ConnectionHandler {
+  case class HandledConnection[F[_], K, M](
+      key: K,
+      serverAddress: InetSocketAddress,
+      underlyingConnection: EncryptedConnection[F, K, M]
+  ) {
+    def sendMessage(m: M): F[Unit] = {
+      underlyingConnection.sendMessage(m)
+    }
+
+    def close(): F[Unit] = {
+      underlyingConnection.close()
+    }
+
+    def incomingMessage: F[Option[Either[ConnectionError, M]]] = {
+      underlyingConnection.incomingMessage
+    }
+  }
+
+  object HandledConnection {
+    def outgoing[F[_], K, M](
+        encryptedConnection: EncryptedConnection[F, K, M]
+    ): HandledConnection[F, K, M] = {
+      HandledConnection(
+        encryptedConnection.remotePeerInfo._1,
+        encryptedConnection.remotePeerInfo._2,
+        encryptedConnection
+      )
+    }
+
+    def incoming[F[_], K, M](
+        serverAddress: InetSocketAddress,
+        encryptedConnection: EncryptedConnection[F, K, M]
+    ): HandledConnection[F, K, M] = {
+      HandledConnection(
+        encryptedConnection.remotePeerInfo._1,
+        serverAddress,
```

Comment:
This is because when we receive incoming connection its remote address is not really remote server address but `remoteIP:ephemeral_port`, so we must take remote server address from some other place. In open p2p network there is discovery or some kinda handshake to exchange this info, here we have closed cluster so we can get peer server address from pre-configured node list

Reply:
Right, this is definitely worth adding to the documentation of `HandledConnection` because I thought the address is the actual one where the connection is from.